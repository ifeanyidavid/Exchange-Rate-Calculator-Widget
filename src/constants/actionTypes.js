export const FETCH_RATES_START = "FETCH_RATES_START";
export const FETCH_RATES_ERROR = "FETCH_RATES_ERROR";
export const FETCH_RATES_DONE = "FETCH_RATES_DONE";
export const UPDATE_SOURCE_CURRENCY = "UPDATE_SOURCE_CURRENCY";
export const UPDATE_DESTINATION_CURRENCY = "UPDATE_DESTINATION_CURRENCY";
export const EXCHANGE_CURRENCY = "EXCHANGE_CURRENCY";
