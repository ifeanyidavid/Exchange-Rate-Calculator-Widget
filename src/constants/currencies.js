export const currencies = {
  EUR: { id: 0, value: "EUR" },
  GBP: { id: 1, value: "GBP" },
  USD: { id: 2, value: "USD" }
};
