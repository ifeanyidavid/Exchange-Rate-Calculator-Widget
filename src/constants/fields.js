export const fieldNames = {
  sourceCurrency: "sourceCurrency",
  destinationCurrency: "destinationCurrency",
  sourceAmount: "sourceAmount",
  destinationAmount: "destinationAmount",
  balance: "balance",
  symbol: "symbol",
  description: "description",
  currency: "currency",
  id: "id",
  error: "error",
  loading: "loading",
  rates: "rates",
};
