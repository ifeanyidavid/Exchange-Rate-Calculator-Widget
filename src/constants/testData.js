import { currencies } from "./currencies";
import { fieldNames } from "./fields";

export const testData = {
  rates: [
    {
      rates: { USD: 1.1075, GBP: 0.84195 },
      base: "EUR",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 1.1877189857, USD: 1.3153987766 },
      base: "GBP",
      date: "2019-12-09"
    },
    {
      rates: { EUR: 0.9029345372, GBP: 0.7602257336 },
      base: "USD",
      date: "2019-12-09"
    }
  ],
  pocketsReducer: {
    transactionHistory: [
      {
        id: 0,
        description: `Exchanged to EUR`,
        time: new Date().toLocaleTimeString("en-GB"),
        sourceAmount: 23,
        destinationAmount: 42,
        sourceCurrency: "USD",
        destinationCurrency: "EUR"
      }
    ],
    pockets: {
      [currencies.EUR.value]: {
        [fieldNames.id]: currencies.EUR.id,
        [fieldNames.currency]: currencies.EUR.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "€",
        [fieldNames.description]: "Euro"
      },
      [currencies.GBP.value]: {
        [fieldNames.id]: currencies.GBP.id,
        [fieldNames.currency]: currencies.GBP.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "£",
        [fieldNames.description]: "British Pound"
      },
      [currencies.USD.value]: {
        [fieldNames.id]: currencies.USD.id,
        [fieldNames.currency]: currencies.USD.value,
        [fieldNames.balance]: 300,
        [fieldNames.symbol]: "$",
        [fieldNames.description]: "American Dollar"
      }
    }
  },
  sourceAmount: 23,
  destinationAmount: 54,
  sourceCurrency: "USD",
  destinationCurrency: "EUR",
  error: new Error("Network disconnected"),
  slideIndex: 1,
  pocketValues: [
    {
      [fieldNames.id]: currencies.EUR.id,
      [fieldNames.currency]: currencies.EUR.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "€",
      [fieldNames.description]: "Euro"
    },
    {
      [fieldNames.id]: currencies.GBP.id,
      [fieldNames.currency]: currencies.GBP.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "£",
      [fieldNames.description]: "British Pound"
    },
    {
      [fieldNames.id]: currencies.USD.id,
      [fieldNames.currency]: currencies.USD.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "$",
      [fieldNames.description]: "American Dollar"
    }
  ]
};
