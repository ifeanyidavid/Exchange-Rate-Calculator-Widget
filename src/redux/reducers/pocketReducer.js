import v4 from "uuid/v4";
import numeral from "numeral";
import * as types from "../../constants/actionTypes";
import { currencies } from "../../constants/currencies";
import { fieldNames } from "../../constants/fields";

export const initialState = {
  pockets: {
    [currencies.EUR.value]: {
      [fieldNames.id]: currencies.EUR.id,
      [fieldNames.currency]: currencies.EUR.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "€",
      [fieldNames.description]: "Euro"
    },
    [currencies.GBP.value]: {
      [fieldNames.id]: currencies.GBP.id,
      [fieldNames.currency]: currencies.GBP.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "£",
      [fieldNames.description]: "British Pound"
    },
    [currencies.USD.value]: {
      [fieldNames.id]: currencies.USD.id,
      [fieldNames.currency]: currencies.USD.value,
      [fieldNames.balance]: 300,
      [fieldNames.symbol]: "$",
      [fieldNames.description]: "American Dollar"
    }
  },
  transactionHistory: []
};

const exchangeCurrency = (state, action) => {
  const sourceAmount = action.payload.sourceAmount;
  const destinationAmount = action.payload.destinationAmount;
  const sourceCurrency = action.payload.sourceCurrency;
  const destinationCurrency = action.payload.destinationCurrency;

  const pockets = state.pockets;
  const sourcePocket = pockets[sourceCurrency];
  const destinationPocket = pockets[destinationCurrency];

  const updatedSource = {
    [sourceCurrency]: {
      [fieldNames.id]: sourcePocket.id,
      [fieldNames.currency]: sourceCurrency,
      [fieldNames.balance]: numeral(sourcePocket.balance)
        .add(sourceAmount)
        .value(),
      [fieldNames.symbol]: sourcePocket.symbol,
      [fieldNames.description]: sourcePocket.description
    }
  };

  const updatedDestination = {
    [destinationCurrency]: {
      [fieldNames.id]: destinationPocket.id,
      [fieldNames.currency]: destinationCurrency,
      [fieldNames.balance]: numeral(destinationPocket.balance)
        .add(destinationAmount)
        .value(),
      [fieldNames.symbol]: destinationPocket.symbol,
      [fieldNames.description]: destinationPocket.description
    }
  };

  const oldTransactionHistory = state.transactionHistory;

  const incomingTransactionHistory = {
    id: v4(),
    description: `Exchanged to ${destinationCurrency}`,
    time: new Date().toLocaleTimeString("en-GB"),
    sourceAmount: sourceAmount,
    destinationAmount: destinationAmount,
    sourceCurrency: sourceCurrency,
    destinationCurrency: destinationCurrency
  };

  return {
    pockets: Object.assign({}, pockets, updatedSource, updatedDestination),
    transactionHistory: [
      ...oldTransactionHistory,
      ...[incomingTransactionHistory]
    ]
  };
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case types.EXCHANGE_CURRENCY: {
      return exchangeCurrency(state, action);
    }
    default:
      return state;
  }
};

export default reducer;
