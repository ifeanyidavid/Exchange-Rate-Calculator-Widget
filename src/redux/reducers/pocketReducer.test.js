import pocketReducer from "./pocketReducer";
import { initialState } from "./pocketReducer";
import { EXCHANGE_CURRENCY } from "../../constants/actionTypes";
import { testData } from "../../constants/testData";

describe("pockets reducer", () => {
  it("should return the initial state", () => {
    expect(pocketReducer(undefined, {})).toMatchSnapshot();
  });
  it("should handle EXCHANGE_CURRENCY", () => {
    expect(
      pocketReducer(initialState, {
        type: EXCHANGE_CURRENCY,
        payload: {
          sourceAmount: testData.sourceAmount,
          destinationAmount: testData.destinationAmount,
          sourceCurrency: testData.sourceCurrency,
          destinationCurrency: testData.destinationCurrency
        }
      })
    ).toMatchSnapshot();
  });
});
