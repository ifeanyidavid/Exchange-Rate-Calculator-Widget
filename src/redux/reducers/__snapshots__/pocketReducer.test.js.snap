// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pockets reducer should handle EXCHANGE_CURRENCY 1`] = `
Object {
  "pockets": Object {
    "EUR": Object {
      "balance": 354,
      "currency": "EUR",
      "description": "Euro",
      "id": 0,
      "symbol": "€",
    },
    "GBP": Object {
      "balance": 300,
      "currency": "GBP",
      "description": "British Pound",
      "id": 1,
      "symbol": "£",
    },
    "USD": Object {
      "balance": 323,
      "currency": "USD",
      "description": "American Dollar",
      "id": 2,
      "symbol": "$",
    },
  },
  "transactionHistory": Array [
    Object {
      "description": "Exchanged to EUR",
      "destinationAmount": 54,
      "destinationCurrency": "EUR",
      "id": "36178e9b-62f2-4e54-90c5-b8cfa6700003",
      "sourceAmount": 23,
      "sourceCurrency": "USD",
      "time": "2:35:40 PM",
    },
  ],
}
`;

exports[`pockets reducer should return the initial state 1`] = `
Object {
  "pockets": Object {
    "EUR": Object {
      "balance": 300,
      "currency": "EUR",
      "description": "Euro",
      "id": 0,
      "symbol": "€",
    },
    "GBP": Object {
      "balance": 300,
      "currency": "GBP",
      "description": "British Pound",
      "id": 1,
      "symbol": "£",
    },
    "USD": Object {
      "balance": 300,
      "currency": "USD",
      "description": "American Dollar",
      "id": 2,
      "symbol": "$",
    },
  },
  "transactionHistory": Array [],
}
`;
